letsencrypt/client/network2.py

```diff
@@ -0,0 +1,480 @@
+        :param contact: Contact list, as accpeted by `.RegistrationResource`
+        :type contact: `tuple`
+
+        :returns: Registration Resource.
+        :rtype: `.RegistrationResource`
+
+        :raises letsencrypt.client.errors.UnexpectedUpdate:
+
+        """
+        new_reg = messages2.Registration(contact=contact)
+
+        response = self._post(self.new_reg_uri, self._wrap_in_jws(new_reg))
+        assert response.status_code == httplib.CREATED  # TODO: handle errors
+
+        regr = self._regr_from_response(response)
+        if regr.body.key != self.key.public() or regr.body.contact != contact:
+            raise errors.UnexpectedUpdate(regr)
+
+        return regr
+
+    def update_registration(self, regr):
+        """Update registration.
+
+        :pram regr: Registration Resource.
+        :type regr: `.RegistrationResource`
+
+        :returns: Updated Registration Resource.
+        :rtype: `.RegistrationResource`
+
+        """
+        response = self._post(regr.uri, self._wrap_in_jws(regr.body))
+
+        # TODO: Boulder returns httplib.ACCEPTED
+        #assert response.status_code == httplib.OK
+
+        # TODO: Boulder does not set Location or Link on update
+        # (c.f. acme-spec #94)
+
+        updated_regr = self._regr_from_response(
+            response, uri=regr.uri, new_authzr_uri=regr.new_authzr_uri,
+            terms_of_service=regr.terms_of_service)
+        if updated_regr != regr:
+            # TODO: Boulder reregisters with new recoveryToken and new URI
+            raise errors.UnexpectedUpdate(regr)
+        return updated_regr
+
+    def _authzr_from_response(self, response, identifier,
+                              uri=None, new_cert_uri=None):
+        if new_cert_uri is None:
+            try:
+                new_cert_uri = response.links['next']['url']
+            except KeyError:
+                raise errors.NetworkError('"next" link missing')
+
+        authzr = messages2.AuthorizationResource(
+            body=messages2.Authorization.from_json(response.json()),
+            uri=response.headers.get('Location', uri),
+            new_cert_uri=new_cert_uri)
+        if (authzr.body.key != self.key.public()
+                or authzr.body.identifier != identifier):
+            raise errors.UnexpectedUpdate(authzr)
+        return authzr
+
+    def request_challenges(self, identifier, regr):
+        """Request challenges.
+
+        :param identifier: Identifier to be challenged.
+        :type identifier: `.messages2.Identifier`
+
+        :pram regr: Registration resource.
+        :type regr: `.RegistrationResource`
+
+        """
+        new_authz = messages2.Authorization(identifier=identifier)
+        response = self._post(regr.new_authzr_uri, self._wrap_in_jws(new_authz))
+        assert response.status_code == httplib.CREATED  # TODO: handle errors
+        return self._authzr_from_response(response, identifier)
+
+    def request_domain_challenges(self, domain, regr):
```

Comment:
Documentation. (sorry for the nitpicks... just trying to make sure it is easy to know how everything works/it is easy to call the functions correctly)